source/common/tls/utility.cc

```diff
@@ -226,6 +226,62 @@ std::string Utility::generalNameAsString(const GENERAL_NAME* general_name) {
     }
     break;
   }
+  case GEN_OTHERNAME: {
+    ASN1_TYPE* value = general_name->d.otherName->value;
+    if (value == nullptr) {
+      break;
+    }
+    switch (value->type) {
+    case V_ASN1_NULL:
+      break;
+    case V_ASN1_BOOLEAN:
+      san = value->value.boolean ? "true" : "false";
+      break;
+    case V_ASN1_ENUMERATED:
+    case V_ASN1_INTEGER: {
+      BIGNUM san_bn;
+      BN_init(&san_bn);
+      value->type == V_ASN1_ENUMERATED ? ASN1_ENUMERATED_to_BN(value->value.enumerated, &san_bn)
+                                       : ASN1_INTEGER_to_BN(value->value.integer, &san_bn);
+      char* san_char = BN_bn2dec(&san_bn);
+      BN_free(&san_bn);
+      if (san_char != nullptr) {
+        san.assign(san_char);
+        OPENSSL_free(san_char);
+      }
+      break;
+    }
+    case V_ASN1_OBJECT: {
+      char tmp_obj[256]; // OID Max length
+      int obj_len = OBJ_obj2txt(tmp_obj, 256, value->value.object, 1);
+      if (obj_len > 256 || obj_len < 0) {
+        break;
+      }
+      san.assign(tmp_obj);
+      break;
+    }
+    case V_ASN1_BIT_STRING:
+    case V_ASN1_OCTET_STRING:
+    case V_ASN1_PRINTABLESTRING:
+    case V_ASN1_T61STRING:
+    case V_ASN1_IA5STRING:
+    case V_ASN1_GENERALSTRING:
+    case V_ASN1_BMPSTRING:
+    case V_ASN1_UNIVERSALSTRING:
+    case V_ASN1_UTCTIME:
+    case V_ASN1_GENERALIZEDTIME:
+    case V_ASN1_VISIBLESTRING:
+    case V_ASN1_UTF8STRING:
+    case V_ASN1_SET:
+    case V_ASN1_SEQUENCE: {
+      str = value->value.asn1_string;
+      san.assign(reinterpret_cast<const char*>(ASN1_STRING_data(str)), ASN1_STRING_length(str));
+      break;
```

Comment:
I think this block is technically correct (all of these types use the same string byte layout in the struct), but it makes me nervous, both because it's difficult to validate the correctness by code inspection, and the possibility that one of the types changes at some point. It would involve a many more lines of code, but I think it would be easier to ensure correctness if each enum value had it's own code block and used the correct named type in the union (even though all these types are typedef'd to the same type). 

We also need some test coverage for each of these.

Reply:
Adding the changes. Initially, I thought the [SET/SEQUENCE](https://www.openssl.org/docs/man3.0/man3/ASN1_generate_nconf.html) values (non-primitive types that are a list of the other primitive type values) could be handled by people through a custom lua string matcher by users. As I'm working through the tests, I wonder if it is the right way. Any thoughts on if we should make the SAN value extracted from the cert something like a UTF-8 string containing the list separated by a delimiter "," 

```
Example:

Config:
asn1 = SEQUENCE:seq_section

[seq_section]

field1 = INTEGER:500
field2 = UTF8:Second
field3 = BOOLEAN:TRUE
```
> Value extracted: 500, Second, TRUE

 or leave it in the DER encoded form for users to handle with a custom matcher?

Reply:
I think it's best to leave it as DER-encoded and have a custom matcher, until someone comes with a concrete use case for doing something different.

Reply:
Updated it